--- etc/sysctl.conf.o	2013-07-11 14:29:42.000000000 -0400
+++ etc/sysctl.conf	2013-07-11 14:29:01.744139363 -0400
@@ -46,6 +46,3 @@
 
 # Speed up the shutdown process
 kern.shutdown.poweroff_delay=500
-
-# Disable DRI debug messages flooding the console
-hw.dri.debug=0
--- etc/rc.conf.pcbsd.o	2013-07-11 14:28:21.058138392 -0400
+++ etc/rc.conf.pcbsd	2013-07-11 14:28:32.472139498 -0400
@@ -105,8 +105,8 @@
   # Enable pbid
   pbid_enable="YES"
 
-  # Enable the gdm display manager
-  gdm_enable="YES"
+  # Enable the pcdm display manager
+  pcdm_enable="YES"
 
   # Enable BlueTooth
   hcsecd_enable="YES"
@@ -139,7 +139,7 @@
 
   # Enable fast-booting
   fastboot_enable="YES"
-  fastboot_earlyrc="/etc/rc.d/moused /usr/local/etc/rc.d/pefs /usr/local/etc/rc.d/dbus /usr/local/etc/rc.d/hald /usr/local/etc/rc.d/gdm"
+  fastboot_earlyrc="/etc/rc.d/moused /usr/local/etc/rc.d/pefs /usr/local/etc/rc.d/dbus /usr/local/etc/rc.d/hald /usr/local/etc/rc.d/pcdm"
 
   # Enable the Warden
   warden_enable="YES"
