diff --git etc/Makefile etc/Makefile
index b39a46f..57c2eac 100644
--- etc/Makefile
+++ etc/Makefile
@@ -25,6 +25,7 @@ BIN1=	crontab \
 	login.access \
 	login.conf \
 	mac.conf \
+	make.conf \
 	motd \
 	netconfig \
 	network.subr \
@@ -36,6 +37,8 @@ BIN1=	crontab \
 	protocols \
 	rc \
 	rc.bsdextended \
+	rc.conf.pcbsd \
+	rc.delay \
 	rc.firewall \
 	rc.initdiskless \
 	rc.sendmail \
diff --git a/etc/csh.cshrc b/etc/csh.cshrc
index f4f775a..c2d0c0f 100644
--- etc/csh.cshrc
+++ etc/csh.cshrc
@@ -1,3 +1,19 @@
-# $FreeBSD$
+# $FreeBSD: src/etc/csh.cshrc,v 1.3 1999/08/27 23:23:40 peter Exp $
 #
 # System-wide .cshrc file for csh(1).
+if ( $?LANG ) then
+else
+  setenv LANG en_US.UTF-8
+endif
+#setenv LC_CTYPE en_US.UTF-8
+#setenv LC_COLLATE POSIX
+setenv  EDITOR  	ee
+setenv  PAGER   	less
+setenv  BLOCKSIZE       K
+
+# Set the MANPATH
+unsetenv MANPATH
+setenv MANPATH `manpath`
+
+# Enable the delete key
+bindkey [3~ delete-char
diff --git a/etc/defaults/periodic.conf b/etc/defaults/periodic.conf
index 9c37c07..37aad91 100644
--- etc/defaults/periodic.conf
+++ etc/defaults/periodic.conf
@@ -230,7 +230,7 @@ weekly_show_info="YES"					# scripts returning 1
 weekly_show_badconfig="NO"				# scripts returning 2
 
 # 310.locate
-weekly_locate_enable="YES"				# Update locate weekly
+weekly_locate_enable="NO"				# Update locate weekly
 
 # 320.whatis
 weekly_whatis_enable="YES"				# Update whatis weekly
diff --git a/etc/defaults/rc.conf b/etc/defaults/rc.conf
index 66022c1..c41a970 100644
--- etc/defaults/rc.conf
+++ etc/defaults/rc.conf
@@ -56,7 +56,7 @@ populate_var="AUTO"	# Set to YES to always (re)populate /var, NO to never
 cleanvar_enable="YES" 	# Clean the /var directory
 local_startup="/usr/local/etc/rc.d" # startup script dirs.
 script_name_sep=" "	# Change if your startup scripts' names contain spaces
-rc_conf_files="/etc/rc.conf /etc/rc.conf.local"
+rc_conf_files="/etc/rc.conf.pcbsd /etc/rc.conf /etc/rc.conf.local"
 
 # ZFS support
 zfs_enable="NO"		# Set to YES to automatically mount ZFS file systems
diff --git a/etc/freebsd-update.conf b/etc/freebsd-update.conf
index 4410903..b769fa4 100644
--- etc/freebsd-update.conf
+++ etc/freebsd-update.conf
@@ -3,23 +3,23 @@
 # Trusted keyprint.  Changing this is a Bad Idea unless you've received
 # a PGP-signed email from <security-officer@FreeBSD.org> telling you to
 # change it and explaining why.
-KeyPrint 800651ef4b4c71c27e60786d7b487188970f4b4169cc055784e21eb71d410cc5
+KeyPrint 771f0e6c8d9de2df1f70a2872a3d25716e6a7a50c32fc4f7ba15746cbf18ce7f
 
 # Server or server pool from which to fetch updates.  You can change
 # this to point at a specific server if you want, but in most cases
 # using a "nearby" server won't provide a measurable improvement in
 # performance.
-ServerName update.FreeBSD.org
+ServerName fbsd-update.pcbsd.org
 
 # Components of the base system which should be kept updated.
-Components src world kernel
+Components world kernel
 
 # Example for updating the userland and the kernel source code only:
 # Components src/base src/sys world
 
 # Paths which start with anything matching an entry in an IgnorePaths
 # statement will be ignored.
-IgnorePaths
+IgnorePaths /boot/kernel/linker.hints
 
 # Paths which start with anything matching an entry in an IDSIgnorePaths
 # statement will be ignored by "freebsd-update IDS".
diff --git a/etc/make.conf b/etc/make.conf
new file mode 100644
index 0000000..83ed646
--- /dev/null
+++ etc/make.conf
@@ -0,0 +1,8 @@
+# Uncomment this if you want to do port builds with no interaction
+#BATCH=yes
+
+# Keep KDE4 in /usr/local, fixes sharing of icons / mime and others
+KDE4_PREFIX=/usr/local
+
+# Enable PKGNG
+WITH_PKGNG=yes
diff --git a/etc/motd b/etc/motd
index cbe55b8..f4fd54f 100644
--- etc/motd
+++ etc/motd
@@ -1,25 +1,3 @@
-FreeBSD ?.?.?  (UNKNOWN)
-
-Welcome to FreeBSD!
-
-Before seeking technical support, please use the following resources:
-
-o  Security advisories and updated errata information for all releases are
-   at http://www.FreeBSD.org/releases/ - always consult the ERRATA section
-   for your release first as it's updated frequently.
-
-o  The Handbook and FAQ documents are at http://www.FreeBSD.org/ and,
-   along with the mailing lists, can be searched by going to
-   http://www.FreeBSD.org/search/.  If the doc package has been installed
-   (or fetched via pkg_add -r lang-freebsd-doc, where lang is the
-   2-letter language code, e.g. en), they are also available formatted
-   in /usr/local/share/doc/freebsd.
-
-If you still have a question or problem, please take the output of
-`uname -a', along with any relevant error messages, and email it
-as a question to the questions@FreeBSD.org mailing list.  If you are
-unfamiliar with FreeBSD's directory layout, please refer to the hier(7)
-manual page.  If you are not familiar with manual pages, type `man man'.
-
-Edit /etc/motd to change this login announcement.
+Welcome to PC-BSD!
 
+For information on the web, visit us at www.pcbsd.org
diff --git a/etc/profile b/etc/profile
index 23c993c..0296d44 100644
--- etc/profile
+++ etc/profile
@@ -1,4 +1,4 @@
-# $FreeBSD$
+# $FreeBSD: src/etc/profile,v 1.14 2004/06/06 11:46:27 schweikh Exp $
 #
 # System-wide .profile file for sh(1).
 #
@@ -12,7 +12,16 @@
 # You should also read the setlocale(3) man page for information
 # on how to achieve more precise control of locale settings.
 #
-# Check system messages
-# msgs -q
+# Read system messages
+# msgs -f
 # Allow terminal messages
 # mesg y
+if [ -z "$LANG" -o "$LANG" = "C" ] ; then
+  LANG="en_US.UTF-8"; export LANG
+fi
+#LC_CTYPE="en_US.UTF-8"; export LC_CTYPE
+#LC_COLLATE="POSIX"; export LC_COLLATE
+EDITOR=ee;      export EDITOR
+PAGER=less;     export PAGER
+BLOCKSIZE=K;    export BLOCKSIZE
+
diff --git a/etc/rc b/etc/rc
index 99cfda8..ccb96a5 100644
--- etc/rc
+++ etc/rc
@@ -96,22 +96,69 @@ done
 
 unset files local_rc
 
-# Now that disks are mounted, for each dir in $local_startup
-# search for init scripts that use the new rc.d semantics.
-#
-case ${local_startup} in
-[Nn][Oo] | '') ;;
-*)	find_local_scripts_new ;;
-esac
+# Check if we are needing to finish up a freebsd-update
+if [ -e "/var/.freebsd-update-finish" ] ; then
+   TERM="cons25" ; export TERM
+   clear
+   echo "Installing FreeBSD updates... Please wait..."
+   /usr/sbin/freebsd-update --non-interactive install
+   if [ -e "/usr/local/bin/pc-softwaremanager" ] ; then
+     /usr/local/bin/pc-extractoverlay desktop
+   else
+     /usr/local/bin/pc-extractoverlay server
+   fi
+   echo "Press ENTER to continue"
+   read tmp
+fi
 
-files=`rcorder ${skip} /etc/rc.d/* ${local_rc} 2>/dev/null`
-for _rc_elem in ${files}; do
+if [ "$fastboot_enable" = "YES" ]; then
+  # We are fast-booting
+
+  # List of rc.d scripts we will start before login
+  preLoad="/etc/rc.d/hostname /etc/rc.d/random /etc/rc.d/adjkerntz /etc/rc.d/cleanvar /etc/rc.d/hostid /etc/rc.d/ldconfig /etc/rc.d/zfs /etc/rc.d/zvol /etc/rc.d/cleartmp $fastboot_earlyrc "
+
+  # Start loopback device
+  /etc/rc.d/netif start lo0
+
+  # Run the early scripts now
+  for i in $preLoad
+  do
+    run_rc_script ${i} ${_boot}
+  done
+
+  # Check for USB mouse to start
+  mChk="ums0 ums1"
+  for i in $mChk
+  do
+    if [ -e "/dev/$i" ] ; then
+       /etc/rc.d/moused start $i
+    fi
+  done
+
+  # Start the delayed services after the login is up
+  (sleep 5; nice /bin/sh /etc/rc.delay "$_rc_elem_done$preLoad")&
+else
+  # Performing a normal startup
+
+  # Now that disks are mounted, for each dir in $local_startup
+  # search for init scripts that use the new rc.d semantics.
+  #
+
+  case ${local_startup} in
+  [Nn][Oo] | '') ;;
+  *)	find_local_scripts_new ;;
+  esac
+
+  files=`rcorder ${skip} /etc/rc.d/* ${local_rc} 2>/dev/null`
+  for _rc_elem in ${files}; do
 	case "$_rc_elem_done" in
 	*" $_rc_elem "*)	continue ;;
 	esac
 
 	run_rc_script ${_rc_elem} ${_boot}
-done
+  done
+
+fi
 
 echo ''
 date
diff --git a/etc/rc.conf.pcbsd b/etc/rc.conf.pcbsd
new file mode 100644
index 0000000..56174bd
--- /dev/null
+++ etc/rc.conf.pcbsd
@@ -0,0 +1,171 @@
+# PC-BSD rc.conf file
+#
+# WARNING - WARNING - WARNING - WARNING - WARNING - WARNING
+#
+# This file *will* be overwritten during upgrades
+# Place your local changes & overrides into /etc/rc.conf
+#
+# WARNING - WARNING - WARNING - WARNING - WARNING - WARNING
+###########################################################
+
+# Defaults for TrueOS servers
+if [ -e "/etc/defaults/trueos" ] ; then
+
+  # FSCK Enhancements
+  fsck_y_enable="YES"
+
+  # Enable pbid
+  pbid_enable="YES"
+
+  # Enable the Warden
+  warden_enable="YES"
+
+  # Load the following kernel modules
+  kld_list="libiconv libmchain msdosfs_iconv sem ntfs ntfs_iconv udf udf_iconv ext2fs pefs xfs reiserfs smbfs fdescfs linsysfs"
+
+  # Enable FSC
+  fscd_enable="YES"
+
+  # Enable ntpd
+  ntpd_enable="YES"
+  ntpd_sync_on_start="YES"
+
+  # Enable IPFW for Warden
+  firewall_enable="YES"
+  firewall_type="open"
+
+  # Build linker.hints files with kldxref(8)
+  # Overwrite old linker.hints at boot
+  kldxref_enable="YES"
+  kldxref_clobber="YES"
+fi
+
+# Defaults for PC-BSD desktops
+if [ -e "/etc/defaults/pcbsd" ] ; then
+
+  background_dhclient="YES" 
+  hostname="pcbsd"
+  compat5x_enable="YES" 
+  sshd_enable="NO" 
+  devd_enable="YES" 
+  devfs_system_ruleset="devfsrules_common" 
+  ldconfig_paths="/usr/lib/compat /usr/local/lib /usr/local/kde4/lib /usr/local/lib/compat/pkg"
+
+  # Disable Sendmail by default
+  sendmail_enable="NONE"
+  sendmail_submit_enable="NO"
+  sendmail_outbound_enable="NO"
+  sendmail_msp_queue_enable="NO" 
+
+  #Enable samba server 
+  samba_enable="YES" 
+  winbindd_enable="YES"
+ 
+  # Disable LPD  
+  lpd_enable="NO" 
+ 
+  # Enable CUPS 
+  cupsd_enable="YES" 
+  linux_enable="YES"
+ 
+  # FSCK Enhancements 
+  fsck_y_enable="YES" 
+ 
+  # powerd: adaptive speed while on AC power, adaptive while on battery power
+  powerd_enable="YES"
+  powerd_flags="-a hiadaptive -b adaptive" # set CPU frequency
+
+  # enable HAL / DBUS
+  dbus_enable="YES"
+  hald_enable="YES"
+
+  # Enable the firewall
+  pf_rules="/etc/pf.conf"
+  pf_enable="YES"
+  pf_flags=""
+
+  # Enable sound-support
+  mixer_enable="YES"
+
+  # Enable avahi_daemon
+  avahi_daemon_enable="YES"
+
+  # Start the swapexd daemon
+  swapexd_enable="YES"
+
+  # Enable IPV6 support
+  ipv6_activate_all_interfaces="YES"
+
+  # Enable BSDStats
+  bsdstats_enable="YES"
+
+  # Enable webcamd
+  webcamd_enable="YES"
+
+  # Disable writing syslogs to remote devices by default
+  syslogd_flags="-c -ss"
+
+  # Enable pbid
+  pbid_enable="YES"
+
+  # Enable the gdm display manager
+  gdm_enable="YES"
+
+  # Enable BlueTooth
+  hcsecd_enable="YES"
+  sdpd_enable="YES"
+
+  # Load the following kernel modules
+  kld_list="libiconv libmchain msdosfs_iconv if_bwn if_bwi bwi_v3_ucode bwn_v4_ucode runfw ng_ubt cuse4bsd sem ntfs ntfs_iconv udf udf_iconv ext2fs mmc mmcsd scd geom_uzip pefs xfs reiserfs smbfs fdescfs linsysfs iwn4965fw iwn1000fw iwn5000fw iwn5150fw iwn6000fw iwn6000g2afw iwn6000g2bfw iwn6050fw acpi_video"
+
+  # Clean out temporary files.
+  clear_tmp_enable="YES"
+  clean_tmp_X="YES"
+
+  # Run the pcbsd init service
+  pcbsdinit_enable="YES"
+
+  # Enable anacron
+  anacron_enable="YES"
+
+  # Enable fusefs
+  fusefs_enable="YES"
+
+  # Enable volmand
+  volmand_enable="YES"
+
+  # Enable pefs
+  pefs_enable="YES"
+
+  # Enable moused
+  moused_enable="YES"
+
+  # Enable fast-booting
+  fastboot_enable="YES"
+  fastboot_earlyrc="/etc/rc.d/moused /usr/local/etc/rc.d/pefs /usr/local/etc/rc.d/dbus /usr/local/etc/rc.d/hald /usr/local/etc/rc.d/gdm"
+
+  # Enable the Warden
+  warden_enable="YES"
+
+  # Enable fail2ban
+  fail2ban_enable="YES"
+
+  # Enable FSC
+  fscd_enable="YES"
+
+  # Enable uhidd
+  uhidd_enable="YES"
+
+  # Enable ntpd
+  ntpd_enable="YES"
+  ntpd_sync_on_start="YES"
+
+  # Enable IPFW for Warden
+  firewall_enable="YES"
+  firewall_type="open"
+
+  # Build linker.hints files with kldxref(8)
+  # Overwrite old linker.hints at boot
+  kldxref_enable="YES"
+  kldxref_clobber="YES"
+fi
diff --git a/etc/rc.delay b/etc/rc.delay
new file mode 100644
index 0000000..df61049
--- /dev/null
+++ etc/rc.delay
@@ -0,0 +1,98 @@
+#!/bin/sh
+#
+# Copyright (c) 2000-2004  The FreeBSD Project
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+#	@(#)rc	5.27 (Berkeley) 6/5/91
+# $FreeBSD$
+#
+
+# Delayed system startup script run by /etc/rc after bootup
+# Output and error are redirected to /var/log/rc_delay.log
+
+# Note that almost all of the user-configurable behavior is no longer in
+# this file, but rather in /etc/defaults/rc.conf.  Please check that file
+# first before contemplating any changes here.  If you do need to change
+# this file for some reason, we would like to know about it.
+
+stty status '^T'
+
+# Set shell to ignore SIGINT (2), but not children;
+# shell catches SIGQUIT (3) and returns to single user.
+#
+trap : 2
+trap "echo 'Boot interrupted'; exit 1" 3
+
+HOME=/
+PATH=/sbin:/bin:/usr/sbin:/usr/bin
+export HOME PATH
+
+autoboot=yes
+_boot="faststart"
+rc_fast=yes        # run_rc_command(): do fast booting
+_log="/var/log/rc_delay.log"
+echo "Starting delayed services `date`" > $_log
+
+dlv=`/sbin/sysctl -n vfs.nfs.diskless_valid 2> /dev/null`
+if [ ${dlv:=0} -ne 0 -o -f /etc/diskless ]; then
+	sh /etc/rc.initdiskless
+fi
+
+# Run these after determining whether we are booting diskless in order
+# to minimize the number of files that are needed on a diskless system,
+# and to make the configuration file variables available to rc itself.
+#
+. /etc/rc.subr
+load_rc_config 'XXX'
+
+skip="-s nostart"
+if [ `/sbin/sysctl -n security.jail.jailed` -eq 1 ]; then
+	skip="$skip -s nojail"
+	if [ "$early_late_divider" = "FILESYSTEMS" ]; then
+		early_late_divider=NETWORKING
+	fi
+fi
+
+unset files local_rc
+_rc_elem_done="$1"
+
+# Now that disks are mounted, for each dir in $local_startup
+# search for init scripts that use the new rc.d semantics.
+#
+case ${local_startup} in
+[Nn][Oo] | '') ;;
+*)	find_local_scripts_new ;;
+esac
+
+files=`rcorder ${skip} /etc/rc.d/* ${local_rc} 2>/dev/null`
+for _rc_elem in ${files}; do
+	case "$_rc_elem_done" in
+	*" $_rc_elem "*)	continue ;;
+	esac
+
+	run_rc_script ${_rc_elem} ${_boot} >>${_log}
+done
+
+echo "Finished starting delayed services `date`" >> $_log
+exit 0
diff --git a/etc/rc.resume b/etc/rc.resume
index 4d17190..58cb245 100755
--- etc/rc.resume
+++ etc/rc.resume
@@ -54,9 +54,15 @@ fi
 
 # wpa_supplicant(8) doesn't seem to reassociate during resume.  Uncomment
 # the following to signal it to reassociate.
-# /usr/sbin/wpa_cli reassociate
+ifconfig | grep -q 'wlan[0-9]:' 
+if [ $? -eq 0 ] ; then
+  /usr/sbin/wpa_cli reassociate
+fi
 
 /usr/bin/logger -t $subsystem resumed at `/bin/date +'%Y%m%d %H:%M:%S'`
 /bin/sync && /bin/sync && /bin/sync
 
+# Restart moused to fix suspend
+/etc/rc.d/moused restart
+
 exit 0
diff --git a/etc/sysctl.conf b/etc/sysctl.conf
index 0c9e796..31f00f6 100644
--- etc/sysctl.conf
+++ etc/sysctl.conf
@@ -1,4 +1,4 @@
-# $FreeBSD$
+# $FreeBSD: src/etc/sysctl.conf,v 1.8 2003/03/13 18:43:50 mux Exp $
 #
 #  This file is read when going to multi-user and its contents piped thru
 #  ``sysctl'' to adjust kernel values.  ``man 5 sysctl.conf'' for details.
@@ -7,3 +7,45 @@
 # Uncomment this to prevent users from seeing information about processes that
 # are being run under another UID.
 #security.bsd.see_other_uids=0
+
+# Disable coredump
+kern.coredump=0
+
+# Up the maxfiles to 4x default
+kern.maxfiles=49312
+
+# Allow users to mount CD's
+vfs.usermount=1
+
+# Autodetect the most recent sound card. Uncomment for Digital output / USB
+#hw.snd.default_auto=1
+
+# Allow pinging in jails
+security.jail.allow_raw_sockets=1
+
+# Required for postgres
+security.jail.sysvipc_allowed=1
+
+# Required for smb
+security.jail.mount_allowed=1
+
+# Solves permission problems for some ports
+security.jail.chflags_allowed=1
+
+# Disable the system speaker
+hw.syscons.bell=0
+
+# Tune for desktop usage
+kern.sched.preempt_thresh=224
+
+# Enable for proper lagg failover 
+#net.link.lagg.failover_rx_all=1
+
+# Enable shm_allow_removed
+kern.ipc.shm_allow_removed=1
+
+# Speed up the shutdown process
+kern.shutdown.poweroff_delay=500
+
+# Disable DRI debug messages flooding the console
+hw.dri.debug=0
diff --git a/sys/boot/forth/beastie.4th b/sys/boot/forth/beastie.4th
index 05b414d..8a1f23b 100644
--- sys/boot/forth/beastie.4th
+++ sys/boot/forth/beastie.4th
@@ -128,21 +128,21 @@ variable logoY
 
 	3 + \ beastie adjustment (see `fbsdbw-logo' comments above)
 
-	2dup at-xy ."  [31m```                        [31;1m`[31m" 1+
-	2dup at-xy ." s` `.....---...[31;1m....--.```   -/[31m" 1+
-	2dup at-xy ." +o   .--`         [31;1m/y:`      +.[31m" 1+
-	2dup at-xy ."  yo`:.            [31;1m:o      `+-[31m" 1+
-	2dup at-xy ."   y/               [31;1m-/`   -o/[31m" 1+
-	2dup at-xy ."  .-                  [31;1m::/sy+:.[31m" 1+
-	2dup at-xy ."  /                     [31;1m`--  /[31m" 1+
-	2dup at-xy ." `:                          [31;1m:`[31m" 1+
-	2dup at-xy ." `:                          [31;1m:`[31m" 1+
-	2dup at-xy ."  /                          [31;1m/[31m" 1+
-	2dup at-xy ."  .-                        [31;1m-.[31m" 1+
-	2dup at-xy ."   --                      [31;1m-.[31m" 1+
-	2dup at-xy ."    `:`                  [31;1m`:`" 1+
-	2dup at-xy ."      [31;1m.--             `--." 1+
-	     at-xy ."         .---.....----.[37m"
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	     at-xy ." "
 
  	\ Put the cursor back at the bottom
  	0 25 at-xy
@@ -152,21 +152,21 @@ variable logoY
 
 	3 + \ beastie adjustment (see `fbsdbw-logo' comments above)
 
-	2dup at-xy ."  ```                        `" 1+
-	2dup at-xy ." s` `.....---.......--.```   -/" 1+
-	2dup at-xy ." +o   .--`         /y:`      +." 1+
-	2dup at-xy ."  yo`:.            :o      `+-" 1+
-	2dup at-xy ."   y/               -/`   -o/" 1+
-	2dup at-xy ."  .-                  ::/sy+:." 1+
-	2dup at-xy ."  /                     `--  /" 1+
-	2dup at-xy ." `:                          :`" 1+
-	2dup at-xy ." `:                          :`" 1+
-	2dup at-xy ."  /                          /" 1+
-	2dup at-xy ."  .-                        -." 1+
-	2dup at-xy ."   --                      -." 1+
-	2dup at-xy ."    `:`                  `:`" 1+
-	2dup at-xy ."      .--             `--." 1+
-	     at-xy ."         .---.....----."
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	2dup at-xy ." " 1+
+	     at-xy ." "
 
  	\ Put the cursor back at the bottom
  	0 25 at-xy
diff --git a/sys/boot/forth/brand.4th b/sys/boot/forth/brand.4th
index 1532203..fb9675a 100644
--- sys/boot/forth/brand.4th
+++ sys/boot/forth/brand.4th
@@ -35,13 +35,13 @@ variable brandY
 
 : fbsd-logo ( x y -- ) \ "FreeBSD" [wide] logo in B/W (7 rows x 42 columns)
 
-	2dup at-xy ."  ______               ____   _____ _____  " 1+
-	2dup at-xy ." |  ____|             |  _ \ / ____|  __ \ " 1+
-	2dup at-xy ." | |___ _ __ ___  ___ | |_) | (___ | |  | |" 1+
-	2dup at-xy ." |  ___| '__/ _ \/ _ \|  _ < \___ \| |  | |" 1+
-	2dup at-xy ." | |   | | |  __/  __/| |_) |____) | |__| |" 1+
-	2dup at-xy ." | |   | | |    |    ||     |      |      |" 1+
-	     at-xy ." |_|   |_|  \___|\___||____/|_____/|_____/ "
+	2dup at-xy ."  _______                   ____    _____" 1+
+	2dup at-xy ." |__   __|                 / __ \  / ____|" 1+
+	2dup at-xy ."    | | _ __  _   _   ___ | |  | || (___" 1+
+	2dup at-xy ."    | || '__|| | | | / _ \| |  | | \___ \" 1+
+	2dup at-xy ."    | || |   | |_| ||  __/| |__| | ____) |" 1+
+	2dup at-xy ."    |_||_|    \__,_| \___| \____/ |_____/" 1+
+	     at-xy ." "
 
 	\ Put the cursor back at the bottom
 	0 25 at-xy
diff --git a/sys/boot/forth/loader.rc b/sys/boot/forth/loader.rc
index 0f9d37e..5d93aff 100644
--- sys/boot/forth/loader.rc
+++ sys/boot/forth/loader.rc
@@ -10,5 +10,15 @@ start
 \ Tests for password -- executes autoboot first if a password was defined
 check-password
 
-\ Unless set otherwise, autoboot is automatic at this point
+\ Check which menu to load
+s" pcbsd_boot" getenv
+-1 = [if]
+	\ Load in the trueos boot menu
+	s" include /boot/beastie.4th" evaluate
+[else]
+	\ Load in the pcbsd boot menu
+	s" include /boot/pcbsd.4th" evaluate
+[then]
 
+\ Start the boot menu
+beastie-start
diff --git a/sys/boot/forth/menu.4th b/sys/boot/forth/menu.4th
index 6ad016d..28fe2e8 100644
--- sys/boot/forth/menu.4th
+++ sys/boot/forth/menu.4th
@@ -417,7 +417,7 @@ create init_text8 255 allot
 
 	\ Print the frame caption at (x,y)
 	s" loader_menu_title" getenv dup -1 = if
-		drop s" Welcome to FreeBSD"
+		drop s" Welcome to TrueOS"
 	then
 	24 over 2 / - 9 at-xy type 
 
diff --git a/sys/boot/forth/pcbsd.4th b/sys/boot/forth/pcbsd.4th
new file mode 100644
index 0000000..12b9811
--- /dev/null
+++ sys/boot/forth/pcbsd.4th
@@ -0,0 +1,465 @@
+\ Copyright (c) 2003 Scott Long <scottl@freebsd.org>
+\ Copyright (c) 2003 Aleksander Fafula <alex@fafula.com>
+\ All rights reserved.
+\
+\ Redistribution and use in source and binary forms, with or without
+\ modification, are permitted provided that the following conditions
+\ are met:
+\ 1. Redistributions of source code must retain the above copyright
+\    notice, this list of conditions and the following disclaimer.
+\ 2. Redistributions in binary form must reproduce the above copyright
+\    notice, this list of conditions and the following disclaimer in the
+\    documentation and/or other materials provided with the distribution.
+\
+\ THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+\ ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+\ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+\ ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+\ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+\ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+\ OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+\ HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+\ LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+\ OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+\ SUCH DAMAGE.
+\
+\ $FreeBSD: src/sys/boot/forth/beastie.4th,v 1.10.2.2 2006/04/04 17:03:44 emax Exp $
+
+marker task-beastie.4th
+
+include /boot/screen.4th
+include /boot/frames.4th
+
+hide
+
+variable menuidx
+variable menubllt
+variable menuX
+variable menuY
+variable promptwidth
+
+variable bootkey
+variable bootacpikey
+variable bootsafekey
+variable bootverbosekey
+variable rundisplaywiz
+variable bootsinglekey
+variable escapekey
+variable rebootkey
+variable bootsafe_set
+variable xvesa
+variable xvesa_set
+variable xwiz_set
+variable kernelkey
+variable kerncurrent
+variable kerncurrent_set
+variable kernold
+variable kernold_set
+variable kerngeneric
+variable kerngeneric_set
+
+46 constant dot
+
+\ The BSD Daemon.  He is 19 rows high and 34 columns wide
+: beastie-logo ( x y -- )
+\ 2dup at-xy ."               [1;31m,        ," 1+
+\ 2dup at-xy ."              /(        )`" 1+
+\ 2dup at-xy ."              \ \___   / |" 1+
+\ 2dup at-xy ."              /- [37m_[31m  `-/  '" 1+
+\ 2dup at-xy ."             ([37m/\/ \[31m \   /\" 1+
+\ 2dup at-xy ."             [37m/ /   |[31m `    \" 1+
+\ 2dup at-xy ."             [34mO O   [37m) [31m/    |" 1+
+\ 2dup at-xy ."             [37m`-^--'[31m`<     '" 1+
+\ 2dup at-xy ."            (_.)  _  )   /" 1+
+\ 2dup at-xy ."             `.___/`    /       " 1+
+\ 2dup at-xy ."               `-----' /" 1+
+\ 2dup at-xy ."  [33m<----.[31m     __ / __   \" 1+
+\ 2dup at-xy ."  [33m<----|====[31mO)))[33m==[31m) \) /[33m====|" 1+
+\ 2dup at-xy ."  [33m<----'[31m    `--' `.__,' \" 1+
+\ 2dup at-xy ."               |        |" 1+
+\ 2dup at-xy ."                \       /       /\" 1+
+\ 2dup at-xy ."           [36m______[31m( (_  / \______/" 1+
+\ 2dup at-xy ."         [36m,'  ,-----'   |" 1+
+\ at-xy ."         `--{__________) [0m"
+;
+
+: beastiebw-logo ( x y -- )
+\	2dup at-xy ."              ,        ," 1+
+\	2dup at-xy ."             /(        )`" 1+
+\	2dup at-xy ."             \ \___   / |" 1+
+\	2dup at-xy ."             /- _  `-/  '" 1+
+\	2dup at-xy ."            (/\/ \ \   /\" 1+
+\	2dup at-xy ."            / /   | `    \" 1+
+\	2dup at-xy ."            O O   ) /    |" 1+
+\	2dup at-xy ."            `-^--'`<     '" 1+
+\	2dup at-xy ."           (_.)  _  )   /" 1+
+\	2dup at-xy ."            `.___/`    /" 1+
+\	2dup at-xy ."              `-----' /" 1+
+\	2dup at-xy ." <----.     __ / __   \" 1+
+\	2dup at-xy ." <----|====O)))==) \) /====" 1+
+\	2dup at-xy ." <----'    `--' `.__,' \" 1+
+\	2dup at-xy ."              |        |" 1+
+\	2dup at-xy ."               \       /       /\" 1+
+\	2dup at-xy ."          ______( (_  / \______/" 1+
+\	2dup at-xy ."        ,'  ,-----'   |" 1+
+\	     at-xy ."        `--{__________)"
+;
+
+: fbsdbw-logo ( x y -- )
+	2dup at-xy ."     ____  ______" 1+
+	2dup at-xy ."    / __ \/ ____/" 1+
+	2dup at-xy ."   / /_/ / /" 1+
+	2dup at-xy ."  / ____/ /___" 1+
+	2dup at-xy ." /_/    \____/" 1+
+	2dup at-xy ."     ____ _____ ____" 1+
+	2dup at-xy ."    / __ ) ___// __ \" 1+
+	2dup at-xy ."   / __  \__ \/ / / /" 1+
+	2dup at-xy ."  / /_/ /__/ / /_/ /" 1+
+	     at-xy ." /_____/____/_____/"
+;
+
+: print-logo ( x y -- )
+	s" loader_logo" getenv
+	dup -1 = if
+		drop
+		fbsdbw-logo
+		exit
+	then
+	2dup s" fbsdbw" compare-insensitive 0= if
+		2drop
+		fbsdbw-logo
+		exit
+	then
+	2dup s" beastiebw" compare-insensitive 0= if
+		2drop
+		beastiebw-logo
+		exit
+	then
+	2dup s" beastie" compare-insensitive 0= if
+		2drop
+		beastie-logo
+		exit
+	then
+	2dup s" none" compare-insensitive 0= if
+		2drop
+		\ no logo
+		exit
+	then
+	2drop
+	fbsdbw-logo
+;
+
+: acpienabled? ( -- flag )
+	s" acpi_load" getenv
+	dup -1 = if
+		drop false exit
+	then
+	s" YES" compare-insensitive 0<> if
+		false exit
+	then
+	s" hint.acpi.0.disabled" getenv
+	dup -1 <> if
+		s" 0" compare 0<> if
+			false exit
+		then
+	else
+		drop
+	then
+	true
+;
+
+: printmenuitem ( -- n )
+	menuidx @
+	1+ dup
+	menuidx !
+	menuY @ + dup menuX @ swap at-xy
+	menuidx @ .
+	menuX @ 1+ swap at-xy
+	menubllt @ emit
+	menuidx @ 48 +
+;
+
+: options-menu ( -- )
+	0 menuidx !
+	dot menubllt !
+	8 menuY !
+	5 menuX !
+	clear
+	24 1 at-xy ."     ____  ______    ____ _____ ____"
+	24 2 at-xy ."    / __ \/ ____/   / __ ) ___// __ \"
+	24 3 at-xy ."   / /_/ / /       / __  \__ \/ / / /"
+	24 4 at-xy ."  / ____/ /___    / /_/ /__/ / /_/ /"
+	24 5 at-xy ." /_/    \____/   /_____/____/_____/"
+	24 6 at-xy ."  "
+	24 7 at-xy ."   Press any key for boot options"
+;
+
+: kernels-menu ( -- )
+	0 menuidx !
+	dot menubllt !
+	8 menuY !
+	5 menuX !
+	clear
+	46 4 print-logo
+	43 15 2 2 box
+	13 6 at-xy ." Select kernel to boot in"
+	printmenuitem ."  Current kernel        (kernel)" kerncurrent !
+	s" kernelold_show" getenv
+	dup -1 <> if
+		s" YES" compare-insensitive 0 = if
+			printmenuitem ."  Previous kernel       (kernel.old)" kernold !
+		then
+	else
+		drop
+	then
+	s" kernelgeneric_show" getenv
+	dup -1 <> if
+		s" YES" compare-insensitive 0 = if
+			printmenuitem ."  Generic PC-BSD kernel (kernel.pcbsd)" kerngeneric !
+		then
+	else
+		drop
+	then
+	menuX @ 15 at-xy
+	." Select option, [Enter] for go back"
+	menuX @ 21 at-xy
+;
+
+: beastie-menu ( -- )
+	-1 menuidx !
+	dot menubllt !
+	8 menuY !
+	5 menuX !
+	clear
+	46 4 print-logo
+	43 22 2 2 box
+	13 5 at-xy ." Welcome to PC-BSD!"
+	printmenuitem ."  Boot PC-BSD [default]" bootkey !
+	printmenuitem ."  Boot PC-BSD in single user mode" bootsinglekey !
+	printmenuitem ."  Boot PC-BSD with verbose logging" bootverbosekey !
+	s" arch-i386" environment? if
+		drop
+		printmenuitem ."  Toggle ACPI " bootacpikey !
+		acpienabled? if
+			." (Enabled)"
+		else
+			." (Disabled)"
+		then
+	else
+		-2 bootacpikey !
+	then
+	printmenuitem ."  Toggle Safe Mode " bootsafekey !
+	bootsafe_set @ 1 = if
+		." (Enabled)"
+	else
+		." (Disabled)"
+	then
+	printmenuitem ."  Toggle X in VESA mode " xvesa !
+	xvesa_set @ 1 = if
+		." (Enabled)"
+	else
+		." (Disabled)"
+	then
+	printmenuitem ."  Run the Display Wizard " rundisplaywiz !
+	xwiz_set @ 1 = if
+		." (Enabled)"
+	else
+		." (Disabled)"
+	then
+	printmenuitem ."  Escape to loader prompt" escapekey !
+	s" kernelmenu_show" getenv
+	dup -1 <> if
+		s" YES" compare-insensitive 0 = if
+			printmenuitem ."  Select kernel " kernelkey !
+			kerngeneric_set @ 1 = if
+				." (kernel.pcbsd)"
+			then
+			kernold_set @ 1 = if
+				." (kernel.old)"
+			then
+		then
+	else
+		drop
+	then
+	printmenuitem ."  Reboot" rebootkey !
+	menuX @ 20 at-xy
+	." Select option, [Enter] for default"
+	menuX @ 21 at-xy
+;
+
+: tkeynotimer
+	seconds +
+	begin 1 while
+		key? if
+			drop
+			key
+			exit
+		then
+	50 ms
+	repeat
+;
+
+: tkeyhidetimer
+	seconds +
+	begin 1 while
+		over 0<> if
+			dup seconds u< if
+				drop
+				-1
+				exit
+			then
+		then
+		key? if
+			drop
+			key
+			exit
+		then
+	50 ms
+	repeat
+;
+
+
+set-current
+
+: beastie-start
+	s" beastie_disable" getenv
+	dup -1 <> if
+		s" YES" compare-insensitive 0= if
+			exit
+		then
+	else
+		drop
+	then
+	s" NO" s" xvesa" setenv
+	0 bootsafe_set !
+	0 xvesa_set !
+	0 xwiz_set !
+	1 kerncurrent_set !
+	0 kernold_set !
+	0 kerngeneric_set !
+	s" autoboot_delay" getenv
+	dup -1 = if
+		s" 02" s" autoboot_delay" setenv
+	then
+	s" autoboot_delay" getenv
+	dup -1 = if
+		drop
+		10
+	else
+		0 0 2swap >number drop drop drop
+	then
+	begin
+	s" autoboot_delay" getenv
+	dup -1 <> if
+		s" NO" compare-insensitive 0= if
+			beastie-menu
+			dup tkeynotimer
+			0 25 at-xy
+			dup 32 = if nip 0 swap then
+			dup -1 = if 0 boot then
+			dup 13 = if 0 boot then
+			dup bootkey @ = if 0 boot then
+			dup bootacpikey @ = if
+				acpienabled? if
+					s" acpi_load" unsetenv
+					s" 1" s" hint.acpi.0.disabled" setenv
+					s" 1" s" loader.acpi_disabled_by_user" setenv
+				else
+					s" YES" s" acpi_load" setenv
+					s" 0" s" hint.acpi.0.disabled" setenv
+				then
+			then
+			dup bootsafekey @ = if
+				bootsafe_set @ 1 = if
+					s" 0" s" hw.ata.ata_dma" setenv
+					s" 0" s" hw.ata.atapi_dma" setenv
+					s" 0" s" hw.ata.wc" setenv
+					s" 0" s" hw.eisa_slots" setenv
+					s" 1" s" hint.kbdmux.0.disabled" setenv
+					0 bootsafe_set !
+				else
+					s" 1" s" hw.ata.ata_dma" setenv
+					s" 1" s" hw.ata.atapi_dma" setenv
+					s" 1" s" hw.ata.wc" setenv
+					s" 1" s" hw.eisa_slots" setenv
+					s" 0" s" hint.kbdmux.0.disabled" setenv
+					1 bootsafe_set !
+				then
+			then
+			dup bootverbosekey @ = if
+				s" YES" s" boot_verbose" setenv
+				0 boot
+			then
+			dup xvesa @ = if
+				xvesa_set @ 1 = if
+					s" NO" s" xvesa" setenv
+					0 xvesa_set !
+				else
+					s" YES" s" xvesa" setenv
+					1 xvesa_set !
+				then
+			then
+			dup rundisplaywiz @ = if
+				xwiz_set @ 1 = if
+					s" NO" s" runwiz" setenv
+					0 xwiz_set !
+				else
+					s" YES" s" runwiz" setenv
+					1 xwiz_set !
+				then
+			then
+			dup bootsinglekey @ = if
+				s" YES" s" boot_single" setenv
+				0 boot
+			then
+			dup kernelkey @ = if
+				kernels-menu
+				dup tkeynotimer
+				dup kerncurrent @ = if
+					1 kerncurrent_set !
+					0 kernold_set !
+					0 kerngeneric_set !
+				then
+				dup kerngeneric @ = if
+					0 kerncurrent_set !
+					0 kernold_set !
+					1 kerngeneric_set !
+				then
+				dup kernold @ = if
+					0 kerncurrent_set !
+					1 kernold_set !
+					0 kerngeneric_set !
+				then
+				0 1 unload drop
+				kerncurrent_set @ 1 = if
+					s" kernel" s" kernel" setenv
+				then
+				kerngeneric_set @  1 = if
+					s" kernel.pcbsd" s" kernel" setenv
+				then
+				kernold_set @ 1 = if
+					s" kernel.old" s" kernel" setenv
+				then
+			then
+			dup escapekey @ = if
+				2drop
+				s" NO" s" autoboot_delay" setenv
+				exit
+			then
+			rebootkey @ = if 0 reboot then
+		else
+			options-menu
+			dup tkeyhidetimer
+			dup 32 = if nip 0 swap then
+			clear
+			dup -1 = if 0 boot then
+			s" NO" s" autoboot_delay" setenv
+		then
+	else
+		clear
+		0 boot	
+	then
+	again
+;
+
+previous
diff --git a/sys/boot/i386/loader/Makefile b/sys/boot/i386/loader/Makefile
index b861e46..ff5e1dd 100644
--- sys/boot/i386/loader/Makefile
+++ sys/boot/i386/loader/Makefile
@@ -98,7 +98,7 @@ FILESMODE_${LOADER}= ${BINMODE} -b
 .if !defined(LOADER_ONLY)
 .PATH: ${.CURDIR}/../../forth 
 FILES+=	loader.help loader.4th support.4th loader.conf
-FILES+= screen.4th frames.4th beastie.4th
+FILES+= screen.4th frames.4th beastie.4th pcbsd.4th
 FILES+= brand.4th check-password.4th color.4th delay.4th
 FILES+= menu.4th menu-commands.4th shortcuts.4th version.4th
 FILESDIR_loader.conf=	/boot/defaults
diff --git a/sys/boot/i386/loader/loader.rc b/sys/boot/i386/loader/loader.rc
index 6443f3f..5d93aff 100644
--- sys/boot/i386/loader/loader.rc
+++ sys/boot/i386/loader/loader.rc
@@ -10,8 +10,15 @@ start
 \ Tests for password -- executes autoboot first if a password was defined
 check-password
 
-\ Load in the boot menu
-include /boot/beastie.4th
+\ Check which menu to load
+s" pcbsd_boot" getenv
+-1 = [if]
+	\ Load in the trueos boot menu
+	s" include /boot/beastie.4th" evaluate
+[else]
+	\ Load in the pcbsd boot menu
+	s" include /boot/pcbsd.4th" evaluate
+[then]
 
 \ Start the boot menu
 beastie-start
diff --git a/usr.sbin/freebsd-update/freebsd-update.sh b/usr.sbin/freebsd-update/freebsd-update.sh
index ea4099a..ae8a513 100644
--- usr.sbin/freebsd-update/freebsd-update.sh
+++ usr.sbin/freebsd-update/freebsd-update.sh
@@ -25,13 +25,14 @@
 # IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 
 # $FreeBSD: releng/9.1/usr.sbin/freebsd-update/freebsd-update.sh 226811 2011-10-26 20:01:43Z cperciva $
 
 #### Usage function -- called from command-line handling code.
 
 # Usage instructions.  Options not listed:
-# --debug	-- don't filter output from utilities
-# --no-stats	-- don't show progress statistics while fetching files
+# --debug		-- don't filter output from utilities
+# --non-interactive	-- run in non-interactive mode
+# --no-stats		-- don't show progress statistics while fetching files
 usage () {
 	cat <<EOF
 usage: `basename $0` [options] command ... [path]
@@ -442,6 +443,10 @@ parse_cmdline () {
 			config_VerboseLevel $1 || usage
 			;;
 
+		--non-interactive)
+			NON_INTERACTIVE=yes
+			;;
+
 		# Aliases for "-v debug" and "-v nostats"
 		--debug)
 			config_VerboseLevel debug || usage
@@ -1022,6 +1027,7 @@ fetch_progress () {
 
 # Function for asking the user if everything is ok
 continuep () {
+ 	if [ "$NON_INTERACTIVE" = "yes" ] ; then return; fi
 	while read -p "Does this look reasonable (y/n)? " CONTINUE; do
 		case "${CONTINUE}" in
 		y*)
@@ -2255,7 +2261,7 @@ upgrade_oldall_to_oldnew () {
 }
 
 # Helper for upgrade_merge: Return zero true iff the two files differ only
 # in the contents of their $FreeBSD: releng/9.1/usr.sbin/freebsd-update/freebsd-update.sh 226811 2011-10-26 20:01:43Z cperciva $ tags.
 samef () {
 	X=`sed -E 's/\\$FreeBSD.*\\$/\$FreeBSD\$/' < $1 | ${SHA256}`
 	Y=`sed -E 's/\\$FreeBSD.*\\$/\$FreeBSD\$/' < $2 | ${SHA256}`
@@ -2351,13 +2357,22 @@ upgrade_merge () {
 		# Ask the user to handle any files which didn't merge.
 		while read F; do
 			# If the installed file differs from the version in
 			# the old release only due to $FreeBSD: releng/9.1/usr.sbin/freebsd-update/freebsd-update.sh 226811 2011-10-26 20:01:43Z cperciva $ tag expansion
 			# then just use the version in the new release.
 			if samef merge/old/${F} merge/${OLDRELNUM}/${F}; then
 				cp merge/${RELNUM}/${F} merge/new/${F}
 				continue
 			fi
 
+ 			if [ "$NON_INTERACTIVE" = "yes" ] ; then
+				cat <<-EOF
+The following file could not be merged automatically: ${F}
+Defaulting to the old copy.
+				EOF
+				cp merge/old/${F} merge/new/${F}
+				continue
+			fi
+
 			cat <<-EOF
 
 The following file could not be merged automatically: ${F}
@@ -2803,6 +2818,9 @@ install_files () {
 		# Do we need to ask for a reboot now?
 		if [ -f $1/kernelfirst ] &&
 		    [ -s INDEX-OLD -o -s INDEX-NEW ]; then
+			if [ "$NON_INTERACTIVE" = "yes" ] ; then
+				touch /var/.freebsd-update-finish
+			fi
 			cat <<-EOF
 
 Kernel updates have been installed.  Please reboot and run
@@ -3241,6 +3259,12 @@ fi
 export LC_ALL=C
 
 get_params $@
+
+# Set PAGER to /bin/cat if non-interactive
+if [ "$NON_INTERACTIVE" = "yes" ] ; then
+	PAGER=/bin/cat
+fi
+
 for COMMAND in ${COMMANDS}; do
 	cmd_${COMMAND}
 done
