--- etc/rc.conf.pcbsd.o	2013-06-19 19:31:00.000000000 -0400
+++ etc/rc.conf.pcbsd	2013-06-19 19:30:28.582984143 -0400
@@ -23,9 +23,6 @@
   # Load the following kernel modules
   kld_list="libiconv libmchain msdosfs_iconv sem ntfs ntfs_iconv udf udf_iconv ext2fs pefs xfs reiserfs smbfs fdescfs linsysfs"
 
-  # Enable FSC
-  fscd_enable="YES"
-
   # Enable ntpd
   ntpd_enable="YES"
   ntpd_sync_on_start="YES"
@@ -150,9 +147,6 @@
   # Enable fail2ban
   fail2ban_enable="YES"
 
-  # Enable FSC
-  fscd_enable="YES"
-
   # Enable uhidd
   uhidd_enable="YES"
 
